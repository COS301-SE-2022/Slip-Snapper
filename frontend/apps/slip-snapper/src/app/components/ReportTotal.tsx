import {
    IonButton,
    IonCard,
    IonCardHeader,
    IonCardSubtitle,
    IonCardTitle,
    IonCol,
    IonItem
} from "@ionic/react";
import { generateReportA } from "../../api/apiCall"
import '../theme/reportTotal.css'

type Props = {
    reportData : string[]
}
function ReportTotal({ reportData }: Props) {
    return (
        <IonCol className='item-col'>
            <IonCard color="primary">
                <IonCardHeader>
                    <IonCardTitle>{reportData[0]}</IonCardTitle>
                    <IonCardSubtitle>{reportData[1]}</IonCardSubtitle>
                </IonCardHeader>
                <IonItem color="tertiary">
                    <IonButton
                        fill="solid"
                        title= {reportData[2]}
                        slot="end"
                        color="secondary"
                        onClick={() => {generateReport(reportData[3])}}
                    >
                        Generate Report
                    </IonButton>
                </IonItem>
            </IonCard>
        </IonCol>
    );
}
export default ReportTotal;

function generateReport(type: string) {
    const url = 'http://localhost:55555/api/report/generate?userId=1&period=' + type + '&userName=1';
    generateReportA(url)
        .then(apiResponse => apiResponse.data);
}
