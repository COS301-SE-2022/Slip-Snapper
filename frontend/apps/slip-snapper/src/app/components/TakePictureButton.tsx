import { receipt, camera, list } from 'ionicons/icons';
import { usePhotoGallery } from '../../hooks/usePhotoGallery';
import {
  IonContent,
  IonFab,
  IonFabButton,
  IonFabList,
  IonIcon,
} from '@ionic/react';
import React from 'react';

/**
 * 
 * @returns Take picture button and Tesseract produced text
 */
const TakePictureButton: React.FC = () => {
  return (
    <IonContent>
        <IonFab horizontal ="center" vertical="bottom" slot="fixed" className='takePicture'>
        <IonFabButton color='secondary'>
          <IonIcon icon={receipt}></IonIcon>
        </IonFabButton>
        <IonFabList side="top">
            <IonFabButton>
              <IonIcon icon={camera} onClick={usePhotoGallery}></IonIcon>
            </IonFabButton>
            <IonFabButton routerLink={"/addentry"}>
              <IonIcon icon={list}></IonIcon>
            </IonFabButton>
          </IonFabList>
      </IonFab>
    </IonContent>
  );
};

export default TakePictureButton;
