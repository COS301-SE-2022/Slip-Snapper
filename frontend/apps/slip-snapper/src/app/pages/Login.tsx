import { IonContent, IonHeader, IonItem, IonLabel, IonPage, IonTextarea} from '@ionic/react';
import React, { useState } from 'react';
import './LoginRegister.css';

const Login: React.FC = () => {
    const [text, setText] = useState<string>();
  
    return (
      <IonPage>
      <IonHeader>
  
      </IonHeader>
      <IonContent fullscreen color="primary">
  
          <IonItem className='inputcredentials'  color="primary">
            <IonLabel position="floating">Username</IonLabel>
            <IonTextarea value={text} onIonChange={e => setText(e.detail.value!)}></IonTextarea>
          </IonItem>
          
          <IonItem className='inputcredentials' color="primary">
          <IonLabel position="floating">Password</IonLabel>
          <IonTextarea value={text} onIonChange={e => setText(e.detail.value!)}></IonTextarea>
        </IonItem>


  
      </IonContent>
    </IonPage>
    );
  };
  
  export default Login;