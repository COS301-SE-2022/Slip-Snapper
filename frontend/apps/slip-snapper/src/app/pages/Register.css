.div {
   
   
   
    padding-top: 5%;
    padding-left: 25%;
    font-size: 16px;
    color: #fff;
    display: list-item;
    max-width: 60%;
    border: 2px;
    border-radius: 5px;
    border-color: "secondary";
    margin: 3px;
    text-align: center;
}

.textArea{
    /* max-width: 60%; */
}
.placeHolder{
    width: 25%;
    height: 25%;
  
    padding-left: 35%;

}
.buttonStyles{
    color: rgb(65, 4, 4);
    /* background-color: blue; */
    padding-left: 35%;
}