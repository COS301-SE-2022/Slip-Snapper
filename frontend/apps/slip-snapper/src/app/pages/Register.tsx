import React, { useState } from 'react';
import {  IonPage,  IonItem, IonLabel, IonButton, IonCard, IonInput} from '@ionic/react';
import '../theme/LoginRegister.css';
const Register: React.FC = () => {
  const [nameInput,setNameInput ] = useState<string>();
  const [surnameInput, setSurnameInput] = useState<string>();
  const [userInput, setUserInput] = useState<string>();
  const [passwordInput, setPasswordInput] = useState<string>();

  return (
    <IonPage>
      <div className="header">
        <div className="inner-header flexbox">
          <IonCard color="tertiary" class="logincard">
            <h1>Slip Snapper</h1>
            <svg
              className="logo"
              baseProfile="tiny"
              id="Layer_1"
              xmlns="http://www.w3.org/2000/svg"
              xmlnsXlink="http://www.w3.org/1999/xlink"
              width="40px"
              height="40px"
              viewBox="0 0 17 17"
              version="1.1"
            >
              <path
                fill="#fff"
                d="M2 0v16.902l2.028-2.481 1.503 1.88 1.501-1.875 1.499 1.875 1.5-1.875 1.5 1.875 
              1.499-1.875 1.97 2.46v-16.886h-13zM14 14.036l-0.97-1.211-1.499 1.875-1.5-1.875-1.5 1.875-1.499-1.875-1.501 
              1.875-1.495-1.87-1.036 1.268v-13.098h11v13.036zM10.997 4h-6v-1h6v1zM8.997 8h-4v-1h4v1zM11.978 6h-7v-1h7v1zM5 10h7v1h-7v-1z"
              />
            </svg>

            <IonItem color="tertiary" class="loginitems">
              <IonLabel position="floating">Name</IonLabel>
              <IonInput
                type="text"
                value={nameInput}
                onIonChange={(e) => setNameInput(e.detail.value!)}
                required
              ></IonInput>
            </IonItem>

            <IonItem color="tertiary" class="loginitems">
              <IonLabel position="floating">Surname</IonLabel>
              <IonInput
                type="text"
                value={surnameInput}
                onIonChange={(e) => setSurnameInput(e.detail.value!)}
                required
              ></IonInput>
            </IonItem>

            <IonItem color="tertiary" class="loginitems">
              <IonLabel position="floating">Username</IonLabel>
              <IonInput
                type="text"
                value={userInput}
                onIonChange={(e) => setUserInput(e.detail.value!)}
                required
              ></IonInput>
            </IonItem>

            <IonItem color="tertiary" class="loginitems">
              <IonLabel position="floating">Password</IonLabel>
              <IonInput
                type="password"
                value={passwordInput}
                onIonChange={(e) => setPasswordInput(e.detail.value!)}
                required
              ></IonInput>
            </IonItem>

            <IonItem color="tertiary" text-align="center" class="loginitems">
              <IonButton
                class="LRButtons"
                color="secondary"
                size="large"
                routerLink={'/home'}
              >
                Register
              </IonButton>
            </IonItem>
          </IonCard>
        </div>

        <div>
          <svg
            className="motionwaves"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
            viewBox="0 24 150 28"
            preserveAspectRatio="none"
            shape-rendering="auto"
          >
            <defs>
              <path
                id="gentle-wave"
                d="M-160 44c30 0 58-18 88-18s 58 18 88 18 58-18 88-18 58 18 88 18 v44h-352z"
              />
            </defs>
            <g className="parallax">
              <use
                xlinkHref="#gentle-wave"
                x="48"
                y="0"
                fill="rgba(235,235,235,0.7"
              />
              <use
                xlinkHref="#gentle-wave"
                x="48"
                y="3"
                fill="rgba(235,235,235,0.5)"
              />
              <use
                xlinkHref="#gentle-wave"
                x="48"
                y="5"
                fill="rgba(235,235,235,0.3)"
              />
              <use xlinkHref="#gentle-wave" x="48" y="7" fill="#ebebeb" />
            </g>
          </svg>
        </div>
      </div>

      <div className="content flexbox"></div>
    </IonPage>
  );

  function register(
    name: string,
    surname: string,
    user: string,
    password: string
  ) {
    const fullName = name + ' ' + surname;
    user = 'Jane Doe';
    fetch('http://localhost:55555/api/user/login', {
      method: 'post',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        name: user,
        userName: fullName,
        pass: password,
      }),
    }).then((res) => res.json());
  }
};



export default Register;