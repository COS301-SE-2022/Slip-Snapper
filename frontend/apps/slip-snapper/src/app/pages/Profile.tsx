import {
  IonContent,
  IonHeader,
  IonPage,
  IonTitle,
  IonToolbar,
  IonButtons,
  IonButton,
  IonCard,
  IonCardHeader,
  IonCardTitle,
  IonItem,
  IonAlert,
  IonProgressBar,
  IonIcon,
  IonText,
} from '@ionic/react';
import React, { useEffect, useState } from 'react';
import { NavButtons } from '../components/NavButtons';
import { EditBudgets } from '../components/EditBudgets';
import '../theme/profile.css';
import '../theme/toasts.css';
import { setBudgetA, getProfileData } from "../../api/apiCall"
import Budget from '../components/Budget';
import { UserStats } from '../components/UserStats';
import { create } from 'ionicons/icons';
import { useHistory } from 'react-router-dom';

const Profile: React.FC = () => {
  const [logoutAlert, setLogoutAlert] = useState(false);
  const [weeklyBudgetAlert, setWeeklyBudgetAlert] = useState(false);
  const [monthlyBudgetAlert, setMonthlyBudgetAlert] = useState(false);
  const [userDetails, setUserDetails] = useState({firstname: "", lastname: ""});
  const val = { weekly: 0, monthly: 0 };
  const history = useHistory();

  let totalWeeklySpent = 300;
  let totalMonthlySpent = 500;
  useEffect(() => {
    let user = JSON.parse(localStorage.getItem('user')!)
    if(user==null){
        user = {id: 24}
    }
    setUserDetails(user)
    getProfileData(user.id)
      .then(
        apiResponse => {
          if(typeof(apiResponse.data) !== "string"){
            val.weekly = apiResponse.data.weekly;
            val.monthly = apiResponse.data.monthly;
            totalWeeklySpent = apiResponse.data.weeklyTotal;
            totalMonthlySpent = apiResponse.data.monthlyTotal;
            setWeeklyBudget(val.weekly)
            setMonthlyBudget(val.monthly)
            setProfile(apiResponse.data)
          }
        })
  }, []);
  const [weeklyBudgetValue, setWeeklyBudget] = useState<number>(val.weekly);
  const [monthlyBudgetValue, setMonthlyBudget] = useState<number>(val.monthly);
  const [profile, setProfile] = useState({favouriteStore:{name:"",receipts:[{id:0,total:0}]},weeklyTotal:0, monthlyTotal:0});
  let weeklyBudget: number, monthlyBudget: number
  return(
    <IonPage>
      <IonHeader>
        <IonToolbar color="primary">
          <IonTitle>Profile</IonTitle>
          <IonButtons slot="end">
            <NavButtons />
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent fullscreen>

        <div className="wrapper">
          <IonCard className="card profile" color="primary">
            <IonCardHeader>
              <IonItem className="headings" color="primary">
                <IonCardTitle>User details</IonCardTitle>
              </IonItem>
              <IonItem className="center-items username" color="tertiary">
                <IonText>Name: {userDetails.firstname} {userDetails.lastname}</IonText>
              </IonItem>
              {/* eslint-disable-next-line jsx-a11y/img-redundant-alt */}
              <img className="profilePhoto" src="..\assets\mock-images\profile-picture-sample.jpg" alt="profile-picture" />
            </IonCardHeader>
          </IonCard>

          <IonCard className="card budget" color="primary">
            <IonCardHeader>
              <IonItem className="headings" color="primary">
                <IonCardTitle>Personal Budget</IonCardTitle>
              </IonItem>
              <IonItem id="weekly-budget" className="center-items" color="tertiary">
                <IonIcon data-testid="weekly-budget-icon" className="edit-budget" src={create} onClick={() => setWeeklyBudgetAlert(true)}/>
                <IonText>Weekly: R{weeklyBudgetValue}</IonText>
                <IonProgressBar id='weeklyProgressBar' class='progressBar' slot="end"></IonProgressBar><br />
              </IonItem>
              <IonItem id="monthly-budget" className="center-items" color="tertiary">
                <IonIcon data-testid="monthly-budget-icon" className="edit-budget" src={create} onClick={() => setMonthlyBudgetAlert(true)}/>
                <IonText>Monthly: R{monthlyBudgetValue}</IonText>
                <IonProgressBar id='monthlyProgressBar' class='progressBar' slot="end"></IonProgressBar><br />
              </IonItem>
            </IonCardHeader>

            {/* Weekly Budget */}
            <IonAlert
              isOpen={weeklyBudgetAlert}
              data-testid="weekly-budget-alert"
              onDidDismiss={() => setWeeklyBudgetAlert(false)}
              header={'Change Budget'}
              inputs={[
                {
                  name: 'weeklyBudget',
                  type: 'text',
                  placeholder: 'Insert Weekly Budget'
                }]}

              buttons={[
                {
                  text: 'Cancel'
                },
                {
                  role: 'applyWeeklyBudget',
                  text: 'Apply',
                  handler: (alertData) => {
                    applyToBudget(alertData.weeklyBudget, "");
                    isExceeded()
                  }
                }
              ]}></IonAlert>
              
              {/* Monthly Budget */}
              <IonAlert
              isOpen={monthlyBudgetAlert}
              data-testid="monthly-budget-alert"
              onDidDismiss={() => setMonthlyBudgetAlert(false)}
              header={'Change Budget'}
              inputs={[
                {
                  id: "monthlyBudget",
                  name: 'monthlyBudget',
                  type: 'text',
                  placeholder: 'Insert Monthly Budget'
                }]}

              buttons={[
                {
                  text: 'Cancel'
                },
                {
                  role: 'applyMonthlyBudget',
                  text: 'Apply',
                  handler: (alertData) => {
                    applyToBudget("", alertData.monthlyBudget);
                    isExceeded()
                  }
                }
              ]}></IonAlert>

            <IonCardHeader>
              <IonItem className="headings" color="primary">
                <IonCardTitle>Category Budgets</IonCardTitle>
              </IonItem>
              <Budget />
              <EditBudgets />
            </IonCardHeader>
          </IonCard>
          
          <IonCard className="card favourite" color="primary">
            <IonCardHeader>
              <IonItem className="headings" color="primary">
                <IonCardTitle>Most Frequent Store</IonCardTitle>
              </IonItem>
              <IonItem className="center-items" color="tertiary">
                <IonText data-testid='favoriteStore'>{profile.favouriteStore.name}</IonText>
              </IonItem>
            </IonCardHeader>

            <IonCardHeader>
              <IonItem className="headings" color="primary">
                <IonCardTitle>Recent Receipts</IonCardTitle>
              </IonItem>
              {profile.favouriteStore.receipts.map((item: any, index: number) => {
                return(
                  <IonItem key={index} className="center-items" color="tertiary">
                    <IonText>Receipt #{item.id}: R{item.total}</IonText>
                  </IonItem>
                )
              })}
            </IonCardHeader>
          </IonCard>
        </div>
        <UserStats />
        <IonButton className="logout-button" onClick={() => setLogoutAlert(true)} expand="block" color='medium'>Logout</IonButton>
        <IonAlert
          isOpen={logoutAlert}
          onDidDismiss={() => setLogoutAlert(false)}
          header="Confirm Logout"
          message="Are you sure you want to logout?"
          buttons={[
            'Cancel',
            {
              text: 'Logout',
              cssClass: 'toasts',
              handler: () => {
                Logout();
              }
            },
          ]}
        />
      </IonContent>
    </IonPage>
  );

  function Logout() {
    history.push("/login")
    window.location.reload()
  }

  function applyToBudget(newWeeklyBudget: string, newMonthlyBudget: string) {
    let user = JSON.parse(localStorage.getItem('user')!)
    if(user==null){
        user = {id: 24}
    }
    weeklyBudget = parseFloat(newWeeklyBudget)
    monthlyBudget = parseFloat(newMonthlyBudget)
    if (!isNaN(weeklyBudget)) {
      setWeeklyBudget(weeklyBudget)
    }
    if (!isNaN(monthlyBudget)) {
      setMonthlyBudget(monthlyBudget)
    }

    setBudgetA(user.id, weeklyBudget, monthlyBudget)
  }

  function isExceeded() {
    totalWeeklySpent = profile.weeklyTotal;
    totalMonthlySpent = profile.monthlyTotal;
    const withinWeeklyBudget = totalWeeklySpent / weeklyBudget
    const withinMonthlyBudget = totalMonthlySpent / monthlyBudget

    if (totalWeeklySpent >= weeklyBudget && !isNaN(weeklyBudget)) {
      document.getElementById("weeklyProgressBar")?.setAttribute("color", "danger")
    }
    else if (totalWeeklySpent >= weeklyBudget / 2 && !isNaN(weeklyBudget)) {
      document.getElementById("weeklyProgressBar")?.setAttribute("color", "warning")
    }
    else if (!isNaN(weeklyBudget)) {
      document.getElementById("weeklyProgressBar")?.setAttribute("color", "success")
    }

    if (totalMonthlySpent >= monthlyBudget && !isNaN(monthlyBudget)) {
      document.getElementById("monthlyProgressBar")?.setAttribute("color", "danger")
    }
    else if (totalMonthlySpent >= monthlyBudget / 2 && !isNaN(monthlyBudget)) {
      document.getElementById("monthlyProgressBar")?.setAttribute("color", "warning")
    }
    else if (!isNaN(monthlyBudget)) {
      document.getElementById("monthlyProgressBar")?.setAttribute("color", "success")
    }

    document.getElementById("weeklyProgressBar")?.setAttribute("value", withinWeeklyBudget.toString())
    document.getElementById("monthlyProgressBar")?.setAttribute("value", withinMonthlyBudget.toString())

    if (weeklyBudget === 0) {
      document.getElementById("weeklyProgressBar")?.setAttribute("value", "0")
      document.getElementById("weeklyProgressBar")?.setAttribute("color", "success")
    }

    if (monthlyBudget === 0) {
      document.getElementById("monthlyProgressBar")?.setAttribute("value", "0")
      document.getElementById("monthlyProgressBar")?.setAttribute("color", "success")
    }

  }
};
export default Profile;