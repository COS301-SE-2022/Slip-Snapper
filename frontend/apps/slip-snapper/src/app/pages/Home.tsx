import {
  IonContent,
  IonHeader,
  IonPage,
  IonTitle,
  IonToolbar,
  IonButtons,
  IonCard,
  IonCardHeader,
  IonCardSubtitle,
  IonCardTitle,
  IonItem,
  IonButton
} from '@ionic/react';
import React from 'react';
import '../theme/variables.css'
import TakePictureButton from '../components/TakePictureButton';
import { NavButtons } from '../components/NavButtons';

const Home: React.FC = () => {
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar color="primary">
          <IonTitle>Slip Snapper</IonTitle>
          <IonButtons slot="end">
            <NavButtons/>
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent fullscreen>
      <IonTitle>Recent Reports</IonTitle>
      <IonCard color="primary">
          <IonCardHeader>
            <IonCardSubtitle>10/05/22</IonCardSubtitle>
            <IonCardTitle>Report #11</IonCardTitle>
          </IonCardHeader>

          <IonItem>
            <IonButton fill="outline" slot="end" color="secondary">View</IonButton>
          </IonItem>
        </IonCard>

        <IonCard color="primary">
          <IonCardHeader>
            <IonCardSubtitle>10/05/22</IonCardSubtitle>
            <IonCardTitle>Report #10</IonCardTitle>
          </IonCardHeader>

          <IonItem>
            <IonButton fill="outline" slot="end" color="secondary">View</IonButton>
          </IonItem>
        </IonCard>

        <IonCard color="primary">
          <IonCardHeader>
            <IonCardSubtitle>10/05/22</IonCardSubtitle>
            <IonCardTitle>Report #9</IonCardTitle>
          </IonCardHeader>

          <IonItem>
            <IonButton fill="outline" slot="end" color="secondary">View</IonButton>
          </IonItem>
        </IonCard>

        <TakePictureButton />
      </IonContent>
    </IonPage>
  );
};

export default Home;
