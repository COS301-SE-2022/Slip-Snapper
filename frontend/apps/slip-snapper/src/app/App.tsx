import React from 'react';

/*Page Imports*/
import Home from './pages/Home';
import ViewReports from './pages/ViewReports';
import Profile from './pages/Profile';
import EditItem from './pages/EditItem';
import Register from './pages/Register';
import Login from './pages/Login'
/*Component Imports*/
import { Redirect, Route } from 'react-router-dom';
import { IonApp, IonRouterOutlet } from '@ionic/react';
import { IonReactRouter } from '@ionic/react-router';
import { Menu } from './components/Menu';

/* Core CSS required for Ionic components to work properly */
import '@ionic/react/css/core.css';

/* Basic CSS for apps built with Ionic */
import '@ionic/react/css/normalize.css';
import '@ionic/react/css/structure.css';
import '@ionic/react/css/typography.css';

/* Optional CSS utils that can be commented out */
import '@ionic/react/css/padding.css';
import '@ionic/react/css/float-elements.css';
import '@ionic/react/css/text-alignment.css';
import '@ionic/react/css/text-transformation.css';
import '@ionic/react/css/flex-utils.css';
import '@ionic/react/css/display.css';

/* Theme variables */
import './theme/variables.css';

const App: React.FC = () => (
  <IonApp>
    <IonReactRouter>
      <Menu/>
      <IonRouterOutlet id="main">
      <Route exact path="/" render={() => <Redirect to="/login" />} />
        <Route path="/home" component={Home} exact={true} />
        <Route path="/viewreports" component={ViewReports} exact={true} />
        <Route path="/profile" component={Profile} exact={true} />
        <Route path="/edititem" component={EditItem} exact={true} />
        <Route path="/register" component={Register} exact={true} />
        <Route path="/login" component={Login} exact={true} />
      </IonRouterOutlet>
    </IonReactRouter>
  </IonApp>
);

export default App;
