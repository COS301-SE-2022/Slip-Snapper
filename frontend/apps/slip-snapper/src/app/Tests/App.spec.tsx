import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import TakePictureButton from '../components/TakePictureButton';
import { create } from 'react-test-renderer';
import App from '../App';
import Profile from '../pages/Profile';
import ViewReports from '../pages/ViewReports';
import Login from '../pages/Login';
import { ionFireEvent as fire } from '@ionic/react-test-utils';
import Register from '../pages/Register';
//ReportTotal Imports
import ReportTotal from '../components/ReportTotal';
import { mockTotals } from '../pages/Home'

test('renders without crashing', () => {
  const { baseElement } = render(<App />);
  expect(baseElement).toBeDefined();
});

describe('TakePicture Button component', () => {
  test('Matches the snapshot of the TakePicture button', () => {
    const button = create(<TakePictureButton />);
    expect(button.toJSON()).toMatchSnapshot();
  });
});

describe('Reports', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('Correctly renders the Reports page', () => {
    const Component = render(<ViewReports />);

    expect(Component.getByText('Profile'));
    expect(Component.getByText('Home'));
    expect(Component.getByText('Reports'));
    expect(Component.getByText('Edit Item'));
    expect(Component.getByText('View Reports'));
  });
});

describe('Profile', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('Correctly renders the Profile page', () => {
    const Component = render(<Profile />);

    expect(Component.getAllByText('Profile'));
    expect(Component.getByText('Home'));
    expect(Component.getByText('Reports'));
    expect(Component.getByText('Edit Item'));
    expect(Component.getByText('User details'));
    expect(Component.getByText('Personal Budget'));
  });

  test('Correctly renders user statitics', async () => {
    const Component = render(<Profile />);

    Component.getByTitle('lastWeek').setAttribute("value", "R100");
    Component.getByTitle('thisWeek').setAttribute("value", "R200");
    Component.getByTitle('lastMonth').setAttribute("value", "R300");
    Component.getByTitle('thisMonth').setAttribute("value", "R400");
    Component.getByTitle("favoriteStore").setAttribute("value", "Woolworths");
    Component.getByTitle("favoriteTotal").setAttribute("value", "R599.99");
    Component.getByTitle("categoryName").setAttribute("value", "Food");
    Component.getByTitle("categoryTotal").setAttribute("value", "R699.99");
    Component.getByTitle("storeName").setAttribute("value", "PEP");
    Component.getByTitle("storeTotal").setAttribute("value", "R899.99");

    expect(Component.getByTitle('lastWeek').getAttribute("value")).toBe("R100")
    expect(Component.getByTitle('thisWeek').getAttribute("value")).toBe("R200")
    expect(Component.getByTitle('lastMonth').getAttribute("value")).toBe("R300")
    expect(Component.getByTitle('thisMonth').getAttribute("value")).toBe("R400")
    expect(Component.getByTitle("favoriteStore").getAttribute("value")).toBe("Woolworths")
    expect(Component.getByTitle("favoriteTotal").getAttribute("value")).toBe("R599.99")
    expect(Component.getByTitle("categoryName").getAttribute("value")).toBe("Food")
    expect(Component.getByTitle("categoryTotal").getAttribute("value")).toBe("R699.99")
    expect(Component.getByTitle("storeName").getAttribute("value")).toBe("PEP")
    expect(Component.getByTitle("storeTotal").getAttribute("value")).toBe("R899.99")
  });
});
/**
 * @returns Jests tests for Login Page
 */
describe('Login', () => {
  it('Correctly renders the login page', () => {
    const Component = render(<Login />);

    expect(Component.getByText('Slip Snapper'));
    expect(Component.getByText('Username'));
    expect(Component.getByText('Password'));
    expect(Component.getByText('Login'));
    expect(Component.getByText('Register'));
  });

  test('Correctly renders user input', async () => {
    const Component = render(<Login />);
    const userInput = await Component.findByTitle('usernameInput');
    const passInput = await Component.findByTitle('passwordInput');

    fire.ionChange(userInput, 'Jane');
    expect(Component.findByText('Jane'));

    fire.ionChange(passInput, 'password123');
    expect(Component.findByText('password123'));
  });
});

/**
 * @returns Jests tests for Register Page
 */
describe('Register', () => {
  it('Correctly renders the register page', () => {
    const Component = render(<Register />);

    expect(Component.getByText('Slip Snapper'));
    expect(Component.getByText('Name'));
    expect(Component.getByText('Surname'));
    expect(Component.getByText('Username'));
    expect(Component.getByText('Password'));
    expect(Component.getByText('Register'));
  });

  test('Correctly renders user input', async () => {
    const Component = render(<Register />);
    const nameInput = await Component.findByTitle('name_Input');
    const surnameInput = await Component.findByTitle('surname_Input');
    const usernameInput = await Component.findByTitle('username_Input');
    const passwordInput = await Component.findByTitle('password_Input');

    fire.ionChange(nameInput, 'Jane');
    fire.ionChange(surnameInput, 'Doe');
    fire.ionChange(usernameInput, 'Jane_Doe777');
    fire.ionChange(passwordInput, 'password123');

    expect(Component.findByText('Jane'));
    expect(Component.findByText('Doe'));
    expect(Component.findByText('Jane_Doe777'));
    expect(Component.findByText('password123'));
  });
});

describe('Register', () => {

  test('Should check if function fires for generate report for daily', async () => {
    const { findByTitle } = render(<ReportTotal reportData={[mockTotals[0].timePeriod, mockTotals[0].total, mockTotals[0].title]} />);
    const addButton = await findByTitle('generateDR');
    fireEvent.click(addButton);
  });
  test('Should check if function fires for generate report for weekly', async () => {
    const { findByTitle } = render(<ReportTotal reportData={[mockTotals[1].timePeriod, mockTotals[1].total, mockTotals[1].title]} />);

    const addButton = await findByTitle('generateWR');
    fireEvent.click(addButton);
  });
  test('Should check if function fires for generate report for Monthly', async () => {
    const { findByTitle } = render(<ReportTotal reportData={[mockTotals[2].timePeriod, mockTotals[2].total, mockTotals[2].title]} />);
    const addButton = await findByTitle('generateMR');
    fireEvent.click(addButton);
  });
});
//Example of old Report API tests
  // test('Should check if function fires for generate report for daily', async () => {
  //   const { findByTitle } = render(<ReportTotal />);
  //   const addButton = await findByTitle('generateMR');
  //   fireEvent.click(addButton);
  // });