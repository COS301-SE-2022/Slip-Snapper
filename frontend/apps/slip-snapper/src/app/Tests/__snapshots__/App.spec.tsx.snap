// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakePicture Button component Matches the snapshot of the TakePicture button 1`] = `
<ion-content>
  <ion-fab
    horizontal="center"
    slot="fixed"
    vertical="bottom"
  >
    <ion-fab-button
      onClick={[Function]}
    >
      <ion-icon
        icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Receipt</title><path d='M483.82 32.45a16.28 16.28 0 00-11.23 1.37L448 46.1l-24.8-12.4a16 16 0 00-14.31 0l-25.11 12.41L359 33.7a16 16 0 00-14.36 0L320 46.07l-24.45-12.34a16 16 0 00-14.35-.06L256 46.12l-24.8-12.43a16.05 16.05 0 00-14.33 0L192 46.1l-24.84-12.41a16 16 0 00-19.36 3.94 16.25 16.25 0 00-3.8 10.65V288l.05.05H336a32 32 0 0132 32V424c0 30.93 33.07 56 64 56h12a52 52 0 0052-52V48a16 16 0 00-12.18-15.55zM416 240H288.5c-8.64 0-16.1-6.64-16.48-15.28A16 16 0 01288 208h127.5c8.64 0 16.1 6.64 16.48 15.28A16 16 0 01416 240zm0-80H224.5c-8.64 0-16.1-6.64-16.48-15.28A16 16 0 01224 128h191.5c8.64 0 16.1 6.64 16.48 15.28A16 16 0 01416 160z'/><path d='M336 424v-88a16 16 0 00-16-16H48a32.1 32.1 0 00-32 32.05c0 50.55 5.78 71.57 14.46 87.57C45.19 466.79 71.86 480 112 480h245.68a4 4 0 002.85-6.81C351.07 463.7 336 451 336 424z'/></svg>"
      />
    </ion-fab-button>
    <ion-fab-list
      side="top"
    >
      <ion-fab-button
        onClick={[Function]}
      >
        <ion-icon
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Camera</title><circle cx='256' cy='272' r='64'/><path d='M432 144h-59c-3 0-6.72-1.94-9.62-5l-25.94-40.94a15.52 15.52 0 00-1.37-1.85C327.11 85.76 315 80 302 80h-92c-13 0-25.11 5.76-34.07 16.21a15.52 15.52 0 00-1.37 1.85l-25.94 41c-2.22 2.42-5.34 5-8.62 5v-8a16 16 0 00-16-16h-24a16 16 0 00-16 16v8h-4a48.05 48.05 0 00-48 48V384a48.05 48.05 0 0048 48h352a48.05 48.05 0 0048-48V192a48.05 48.05 0 00-48-48zM256 368a96 96 0 1196-96 96.11 96.11 0 01-96 96z'/></svg>"
          onClick={[Function]}
        />
      </ion-fab-button>
      <ion-fab-button
        href="/addentry"
        onClick={[Function]}
        router-link="/addentry"
      >
        <ion-icon
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>List</title><path stroke-linecap='round' stroke-linejoin='round' stroke-width='48' d='M160 144h288M160 256h288M160 368h288' class='ionicon-fill-none'/><circle cx='80' cy='144' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='80' cy='256' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='80' cy='368' r='16' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
        />
      </ion-fab-button>
    </ion-fab-list>
  </ion-fab>
</ion-content>
`;
